git-remote-ipfs/git-remote-ipfs.cabal

```diff
@@ -106,6 +107,8 @@ executable git-remote-ipfs
       , optparse-applicative
       , text
 
+    ghc-options: -threaded -rtsopts "-with-rtsopts=-maxN4 -A8m"
```

Comment:
I never had to use `-maxN` before (my go-to has always been `-N`) and from the docs I can see how they are subtly different. Quoting [this page](https://downloads.haskell.org/~ghc/master/users-guide/using-concurrent.html) :

- "[...]The runtime manages a set of virtual processors, which we call capabilities, the number of which is determined by the -N option[...] Omitting ⟨x⟩, i.e. +RTS -N -RTS, lets the runtime choose the value of ⟨x⟩ itself based on how many processors are in your machine."

- "With -maxN⟨x⟩, i.e. +RTS -maxN3 -RTS, the runtime will choose at most (x), also limited by the number of processors on the system. Omitting (x) is an error, if you need a default use option -N."

It also warns you from using something ad-hoc : "Be careful when using all the processors in your machine: if some of your processors are in use by other programs, this can actually harm performance rather than improve it. Asking GHC to create more capabilities than you have physical threads is almost always a bad idea."

Having said all of that, if this code has to run of multiple computers, shouldn't we simply use `-N` and let the RTS figure out the actual number of capabilities?

Reply:
`-N` tends to be a bad default for almost all programs due to GC (there are a lot of `-q*` options to twiddle to get this right). In fact, using `-N` for the trivial git histories in the tests makes things 2-3 times slower than not using the threaded runtime at all (which isn't all that surprising).

If I understand you correctly, you're worried that on a dual core machine `-maxN4` will use up all available cores, right? It does _seem_ to do the right thing, but you're right that there isn't a reason why it should. I guess the only way to make this portable is to use `getNumProcessors` (which unfortunately counts HTs as well) and `setNumCapabilities` :disappointed:

Reply:
> -N tends to be a bad default for almost all programs due to GC (there are a lot of -q* options to twiddle to get this right).

Yes, using `-N` also enables the parallel GC and that indeed requires a bit of tuning with those `-q` options, and that's a bit of a manual process as it really depends on the allocation pattern of your application, that's true.

> If I understand you correctly, you're worried that on a dual core machine -maxN4 will use up all available cores, right?

Not so much that, but more the risk of running this on a machine with only 1 core (agreed, they are extremely rare these days) or under-utilise newer machines having 4 or more cores.

Nothing worth of blocking this PR, but perhaps something we might want to capture in a ticket? I can see how we might want to do something similar for `oscoin` as well.

Reply:
#30

> running this on a machine with only 1 core

:thinking: `-maxN<x>` is essentially `min x <$> getNumProcessors`

> under-utilise newer machines

As I said, this doesn't seem to be the case - performance gets worse with > 4 capabilities. I think this makes sense for an event-driven I/O manager like GHC's, also considering that we're waiting for socket I/O which _actually_ benefits from `O_NONBLOCK`.

Reply:
Oh, sorry, I did misinterpret this one below:

> With -maxN⟨x⟩, i.e. +RTS -maxN3 -RTS, the runtime will choose at most (x), also limited by the number of processors on the system. Omitting (x) is an error, if you need a default use option -N.

For some reason, my brain was reading this like "use 3 cores no matter what (see -maxN3 in the example) but if there are more, don't use them. But instead this is saying "if you have less cores, use as much as you have, but if you have more, use at max 3".

All good, sorry for the confusion!